crates/red_knot_python_semantic/resources/mdtest/generics/functions.md

```diff
@@ -0,0 +1,103 @@
+# Generic functions
+
+## Inferring generic function parameter types
+
+If the type of a generic function parameter is a typevar, then we can infer what type that typevar
+is bound to at each call site.
+
+```py
+def f[T](x: T) -> T: ...
+
+# TODO: no error
+# TODO: revealed: int
+# error: [invalid-argument-type]
+reveal_type(f(1))  # revealed: T
+
+# TODO: no error
+# TODO: revealed: float
+# error: [invalid-argument-type]
+reveal_type(f(1.0))  # revealed: T
+
+# TODO: no error
+# TODO: revealed: bool
+# error: [invalid-argument-type]
+reveal_type(f(True))  # revealed: T
+
+# TODO: no error
+# TODO: revealed: str
+# error: [invalid-argument-type]
+reveal_type(f("string"))  # revealed: T
+```
+
+## Inferring “deep” generic parameter types
+
+The matching up of call arguments and discovery of constraints on typevars can be a recursive
+process for arbitrarily-nested generic types in parameters.
+
+```py
+def f[T](x: list[T]) -> T: ...
+
+# TODO: revealed: float
+reveal_type(f([1.0, 2.0]))  # revealed: T
+```
+
+## Return type uses actual typevar, not upper bound
+
+If a function is annotated as returning a typevar, the return type must be an instance of the actual
+typevar, and not just something assignable to its upper bound.
```

Comment:
By "the return type", do you mean "types of `return` expressions" or similar?

I think "must be an instance of the actual typevar" is too narrow? Shouldn't it be something like "if a function is annotated as returning a typevar `T`, types of `return` expressions must be assignable to `T`, and not …"?

Reply:
I think this whole sentence is an attempt to clarify what "assignable to T" means. That is, if `T` has an upper bound of `int`, that does not make `int` assignable to `T`.

Reply:
> I took a stab at rewording this, lmkwyt

I think it's clear now, thank you.